src/components/cylc/tree/tree.js

```diff
@@ -0,0 +1,497 @@
+        this.lookup.delete(n.id)
+        if (n.children && n.children.length > 0) {
+          stack.push(...n.children)
+        }
+      }
+    }
+  }
+
+  // --- Cycle points
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} cyclePoint
+   */
+  addCyclePoint (cyclePoint) {
+    if (cyclePoint && !this.lookup.has(cyclePoint.id)) {
+      this.lookup.set(cyclePoint.id, cyclePoint)
+      this.root.children.push(cyclePoint)
+      // sort cycle points
+      this.root.children.sort((cyclepoint, anotherCyclepoint) => {
+        return cyclepoint.id.localeCompare(anotherCyclepoint.id)
+      })
+    }
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} cyclePoint
+   */
+  updateCyclePoint (cyclePoint) {
+    if (cyclePoint) {
+      const node = this.lookup.get(cyclePoint.id)
+      if (node) {
+        merge(node, cyclePoint)
+      }
+    }
+  }
+
+  /**
+   * @param {null|string} cyclePointId
+   */
+  removeCyclePoint (cyclePointId) {
+    const node = this.lookup.get(cyclePointId)
+    if (node) {
+      this.recursivelyRemoveNode(node)
+      this.root.children.splice(this.root.children.indexOf(node), 1)
+    }
+  }
+
+  tallyCyclePointStates () {
+    // calculate cycle point states
+    computeCyclePointsStates(this.root.children)
+  }
+
+  // --- Family proxies
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} familyProxy
+   */
+  addFamilyProxy (familyProxy) {
+    // When we receive the families from the GraphQL endpoint, we are sorting by their
+    // firstParent's. However, you may get family proxies out of order when iterating
+    // them. When that happens, you may add a family proxy to the lookup, and only
+    // append it to the parent later.
+    if (familyProxy) {
+      // ignore the root family
+      if (familyProxy.id.endsWith(`|${FAMILY_ROOT}`)) {
+        return
+      }
+      // add if not in the lookup already
+      const existingFamilyProxy = this.lookup.get(familyProxy.id)
+      if (!existingFamilyProxy) {
+        this.lookup.set(familyProxy.id, familyProxy)
+      } else {
+        // We may get a family proxy added twice. The first time is when it is the parent of another
+        // family proxy. In that case, we create an orphan node in the lookup table.
+        // The second time will be node with more information, such as .firstParent {}. When this happens,
+        // we must remember to merge the objects.
+        merge(existingFamilyProxy, familyProxy)
+        this.lookup.set(existingFamilyProxy.id, existingFamilyProxy)
+        // NOTE: important, replace the version so that we use the existing one
+        // when linking with the parent node in the tree, not the new GraphQL data
+        familyProxy = existingFamilyProxy
+      }
+      if (!familyProxy.node.state) {
+        familyProxy.node.state = ''
+      }
```

Comment:
Is this to handle ghost nodes?

Might be worth adding a comment here (like the earlier one) incase we handle this at the scheduler in the future.

Reply:
I think that's because `familyProxy`'s don't have states. So the UI (vue template/dom) would have to protect against `undefined`.

I found it easier to just define the state here. Further on, after the tasks/jobs are added, we then compute the family proxy state based on its children states (using old Cylc 7 logic).

(This is in theory, it's been implemented some time ago, and hopefully that's working OK... we have tests for that I think, but before 1.0 release we need to verify that the family statuses are being updated correctly I guess)

Reply:
```
      // See comment above in the else block. When we get family proxies out of order, we create the parent
      // nodes if they don't exist in the tree yet, so that we can create the correct hierarchy. Later, we
      // merge the data of the node. But for a while, the family proxy that we create won't have a state (as
      // the state is given in the deltas data, and is not available in the `node.firstParent { id }`.
```

(pushed in a separate commit)